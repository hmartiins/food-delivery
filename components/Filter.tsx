import { useState } from 'react';

import { FlatList, Platform, Text, TouchableOpacity } from 'react-native';

import { router, useLocalSearchParams } from 'expo-router';

import cn from 'clsx';

import { Category } from '@/type';

export const Filter = ({ categories }: { categories: Category[] }) => {
  const searchParams = useLocalSearchParams();
  const [active, setActive] = useState(searchParams.category || '');

  const handlePress = (id: string) => {
    setActive(id);

    if (id === 'all') {
      router.setParams({ category: undefined });
    } else {
      router.setParams({ category: id });
    }
  };

  const filterData: (Category | { $id: string; name: string })[] = categories
    ? [{ $id: 'all', name: 'All' }, ...categories]
    : [{ $id: 'all', name: 'All' }];

  return (
    <FlatList
      testID="filter-list"
      data={filterData}
      horizontal
      showsHorizontalScrollIndicator={false}
      contentContainerClassName="gap-x-2 pb-3"
      accessible={true}
      accessibilityRole="tablist"
      accessibilityLabel="Category filters"
      accessibilityHint="Swipe horizontally to browse different category filters"
      renderItem={({ item }) => {
        const isActive = active === item.$id;

        return (
          <TouchableOpacity
            key={item.$id}
            testID={`filter-item-${item.$id}`}
            onPress={() => handlePress(item.$id)}
            className={cn('filter', isActive ? 'bg-amber-500' : 'bg-white')}
            style={
              Platform.OS === 'android'
                ? { elevation: 5, shadowColor: '#878787' }
                : {}
            }
            accessible={true}
            accessibilityRole="button"
            accessibilityState={{ selected: isActive }}
            accessibilityLabel={`${item.name} filter`}
            accessibilityHint={
              isActive
                ? `${item.name} filter is currently active`
                : `Tap to filter by ${item.name}`
            }
          >
            <Text
              testID={`filter-text-${item.$id}`}
              className={cn(
                'body-medium',
                isActive ? 'text-white' : 'text-gray-200'
              )}
              accessible={false}
            >
              {item.name}
            </Text>
          </TouchableOpacity>
        );
      }}
      keyExtractor={item => item.$id}
    />
  );
};
