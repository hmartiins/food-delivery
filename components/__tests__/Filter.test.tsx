import { router, useLocalSearchParams } from 'expo-router';

import { fireEvent, render } from '@testing-library/react-native';

import { Category } from '@/type';

import { Filter } from '../Filter';

jest.mock('expo-router', () => ({
  router: { setParams: jest.fn() },
  useLocalSearchParams: jest.fn(),
}));

jest.mock('clsx', () => jest.fn((className: string) => className));

const mockUseLocalSearchParams = useLocalSearchParams as jest.MockedFunction<
  typeof useLocalSearchParams
>;

describe('<Filter />', () => {
  const mockCategories: Category[] = [
    {
      $id: 'cat1',
      name: 'Pizza',
      description: 'Delicious pizza',
      $collectionId: 'collection1',
      $databaseId: 'database1',
      $createdAt: '2024-01-01T00:00:00.000Z',
      $updatedAt: '2024-01-01T00:00:00.000Z',
      $permissions: [],
    },
    {
      $id: 'cat2',
      name: 'Burger',
      description: 'Tasty burgers',
      $collectionId: 'collection1',
      $databaseId: 'database1',
      $createdAt: '2024-01-01T00:00:00.000Z',
      $updatedAt: '2024-01-01T00:00:00.000Z',
      $permissions: [],
    },
    {
      $id: 'cat3',
      name: 'Salad',
      description: 'Fresh salads',
      $collectionId: 'collection1',
      $databaseId: 'database1',
      $createdAt: '2024-01-01T00:00:00.000Z',
      $updatedAt: '2024-01-01T00:00:00.000Z',
      $permissions: [],
    },
  ];

  beforeEach(() => jest.clearAllMocks());

  it('renders correctly with empty categories', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId, getByText } = render(<Filter categories={[]} />);

    expect(getByTestId('filter-list')).toBeTruthy();
    expect(getByTestId('filter-item-all')).toBeTruthy();
    expect(getByText('All')).toBeTruthy();
  });

  it('renders correctly with categories', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId, getByText } = render(
      <Filter categories={mockCategories} />
    );

    expect(getByTestId('filter-list')).toBeTruthy();
    expect(getByTestId('filter-item-all')).toBeTruthy();
    expect(getByTestId('filter-item-cat1')).toBeTruthy();
    expect(getByTestId('filter-item-cat2')).toBeTruthy();
    expect(getByTestId('filter-item-cat3')).toBeTruthy();

    expect(getByText('All')).toBeTruthy();
    expect(getByText(mockCategories[0].name)).toBeTruthy();
    expect(getByText(mockCategories[1].name)).toBeTruthy();
    expect(getByText(mockCategories[2].name)).toBeTruthy();
  });

  it('sets initial active state from search params', () => {
    mockUseLocalSearchParams.mockReturnValue({ category: 'cat1' });

    const { getByTestId } = render(<Filter categories={mockCategories} />);

    const activeItem = getByTestId('filter-item-cat1');
    const inactiveItem = getByTestId('filter-item-all');

    expect(activeItem.props.accessibilityState.selected).toBe(true);
    expect(inactiveItem.props.accessibilityState.selected).toBe(false);
  });

  it('calls router.setParams with undefined when "All" is pressed', () => {
    mockUseLocalSearchParams.mockReturnValue({ category: 'cat1' });

    const { getByTestId } = render(<Filter categories={mockCategories} />);

    const allButton = getByTestId('filter-item-all');
    fireEvent.press(allButton);

    expect(router.setParams).toHaveBeenCalledWith({ category: undefined });
  });

  it('calls router.setParams with category id when category is pressed', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId } = render(<Filter categories={mockCategories} />);

    const pizzaButton = getByTestId('filter-item-cat1');
    fireEvent.press(pizzaButton);

    expect(router.setParams).toHaveBeenCalledWith({ category: 'cat1' });
  });

  it('updates active state when filter item is pressed', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId } = render(<Filter categories={mockCategories} />);

    const burgerButton = getByTestId('filter-item-cat2');
    fireEvent.press(burgerButton);

    expect(router.setParams).toHaveBeenCalledWith({ category: 'cat2' });
  });

  it('handles multiple category presses correctly', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId } = render(<Filter categories={mockCategories} />);

    // Press Pizza
    const pizzaButton = getByTestId('filter-item-cat1');
    fireEvent.press(pizzaButton);
    expect(router.setParams).toHaveBeenCalledWith({ category: 'cat1' });

    // Press Burger
    const burgerButton = getByTestId('filter-item-cat2');
    fireEvent.press(burgerButton);
    expect(router.setParams).toHaveBeenCalledWith({ category: 'cat2' });

    // Press All
    const allButton = getByTestId('filter-item-all');
    fireEvent.press(allButton);
    expect(router.setParams).toHaveBeenCalledWith({ category: undefined });

    expect(router.setParams).toHaveBeenCalledTimes(3);
  });

  describe('Accessibility', () => {
    it('has correct accessibility properties on filter list', () => {
      mockUseLocalSearchParams.mockReturnValue({});

      const { getByTestId } = render(<Filter categories={mockCategories} />);

      const filterList = getByTestId('filter-list');
      expect(filterList.props.accessible).toBe(true);
      expect(filterList.props.accessibilityRole).toBe('tablist');
      expect(filterList.props.accessibilityLabel).toBe('Category filters');
      expect(filterList.props.accessibilityHint).toBe(
        'Swipe horizontally to browse different category filters'
      );
    });

    it('has correct accessibility properties on active filter item', () => {
      mockUseLocalSearchParams.mockReturnValue({ category: 'cat1' });

      const { getByTestId } = render(<Filter categories={mockCategories} />);

      const activeItem = getByTestId('filter-item-cat1');
      expect(activeItem.props.accessible).toBe(true);
      expect(activeItem.props.accessibilityRole).toBe('button');
      expect(activeItem.props.accessibilityState.selected).toBe(true);
      expect(activeItem.props.accessibilityLabel).toBe('Pizza filter');
      expect(activeItem.props.accessibilityHint).toBe(
        'Pizza filter is currently active'
      );
    });

    it('has correct accessibility properties on inactive filter item', () => {
      mockUseLocalSearchParams.mockReturnValue({ category: 'cat1' });

      const { getByTestId } = render(<Filter categories={mockCategories} />);

      const inactiveItem = getByTestId('filter-item-cat2');
      expect(inactiveItem.props.accessible).toBe(true);
      expect(inactiveItem.props.accessibilityRole).toBe('button');
      expect(inactiveItem.props.accessibilityState.selected).toBe(false);
      expect(inactiveItem.props.accessibilityLabel).toBe('Burger filter');
      expect(inactiveItem.props.accessibilityHint).toBe(
        'Tap to filter by Burger'
      );
    });
  });

  it('snapshot test with categories', () => {
    mockUseLocalSearchParams.mockReturnValue({ category: 'cat1' });

    const { toJSON } = render(<Filter categories={mockCategories} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('snapshot test without categories', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { toJSON } = render(<Filter categories={[]} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
