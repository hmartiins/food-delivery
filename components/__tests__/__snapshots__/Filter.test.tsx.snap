// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> snapshot test with categories 1`] = `
<RCTScrollView
  accessibilityHint="Swipe horizontally to browse different category filters"
  accessibilityLabel="Category filters"
  accessibilityRole="tablist"
  accessible={true}
  contentContainerClassName="gap-x-2 pb-3"
  data={
    [
      {
        "$id": "all",
        "name": "All",
      },
      {
        "$collectionId": "collection1",
        "$createdAt": "2024-01-01T00:00:00.000Z",
        "$databaseId": "database1",
        "$id": "cat1",
        "$permissions": [],
        "$updatedAt": "2024-01-01T00:00:00.000Z",
        "description": "Delicious pizza",
        "name": "Pizza",
      },
      {
        "$collectionId": "collection1",
        "$createdAt": "2024-01-01T00:00:00.000Z",
        "$databaseId": "database1",
        "$id": "cat2",
        "$permissions": [],
        "$updatedAt": "2024-01-01T00:00:00.000Z",
        "description": "Tasty burgers",
        "name": "Burger",
      },
      {
        "$collectionId": "collection1",
        "$createdAt": "2024-01-01T00:00:00.000Z",
        "$databaseId": "database1",
        "$id": "cat3",
        "$permissions": [],
        "$updatedAt": "2024-01-01T00:00:00.000Z",
        "description": "Fresh salads",
        "name": "Salad",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={true}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  showsHorizontalScrollIndicator={false}
  stickyHeaderIndices={[]}
  testID="filter-list"
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityHint="Tap to filter by All"
        accessibilityLabel="All filter"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="filter-item-all"
      >
        <Text
          accessible={false}
          className="body-medium"
          testID="filter-text-all"
        >
          All
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityHint="Pizza filter is currently active"
        accessibilityLabel="Pizza filter"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": true,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="filter-item-cat1"
      >
        <Text
          accessible={false}
          className="body-medium"
          testID="filter-text-cat1"
        >
          Pizza
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityHint="Tap to filter by Burger"
        accessibilityLabel="Burger filter"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="filter-item-cat2"
      >
        <Text
          accessible={false}
          className="body-medium"
          testID="filter-text-cat2"
        >
          Burger
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityHint="Tap to filter by Salad"
        accessibilityLabel="Salad filter"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="filter-item-cat3"
      >
        <Text
          accessible={false}
          className="body-medium"
          testID="filter-text-cat3"
        >
          Salad
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`<Filter /> snapshot test without categories 1`] = `
<RCTScrollView
  accessibilityHint="Swipe horizontally to browse different category filters"
  accessibilityLabel="Category filters"
  accessibilityRole="tablist"
  accessible={true}
  contentContainerClassName="gap-x-2 pb-3"
  data={
    [
      {
        "$id": "all",
        "name": "All",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={true}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  showsHorizontalScrollIndicator={false}
  stickyHeaderIndices={[]}
  testID="filter-list"
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={
        [
          {
            "flexDirection": "row",
          },
          null,
        ]
      }
    >
      <View
        accessibilityHint="Tap to filter by All"
        accessibilityLabel="All filter"
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="filter-item-all"
      >
        <Text
          accessible={false}
          className="body-medium"
          testID="filter-text-all"
        >
          All
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
